http/src/main/scala/org/http4s/blaze/http/HttpServerStage.scala

```diff
@@ -1,262 +1,113 @@
-    }
-  }
-
-  private def completionHandler(result: Try[RouteResult]): Unit = result match {
-    case Success(completed) =>
-      completed match {
-      case Close           => shutdownWithCommand(Cmd.Disconnect)
-      case Upgrade         => // NOOP: don't need to do anything
-      case Reload          =>
-        val buff = remainder
-        resetStage()
-        readLoop(buff)
-    }
-    case Failure(t: BadRequest)   => badRequest(t)
-    case Failure(t)               => shutdownWithCommand(Cmd.Error(t))
-  }
-
-  private def getEncoder(forceClose: Boolean, prelude: HttpResponsePrelude): InternalWriter = {
-    val sb = new StringBuilder(512)
-    sb.append("HTTP/").append(1).append('.').append(minor).append(' ')
-      .append(prelude.code).append(' ')
-      .append(prelude.status).append('\r').append('\n')
-
-    val keepAlive = !forceClose && isKeepAlive(prelude.headers)
-
-    if (!keepAlive) sb.append("connection: close\r\n")
-    else if (minor == 0 && keepAlive) sb.append("Connection: Keep-Alive\r\n")
-
-    InternalWriter.selectWriter(!keepAlive, prelude, sb, this)
-  }
-
-  /** Deal with route response of WebSocket form */
-  private def handleWebSocket(reqHeaders: Headers, wsBuilder: LeafBuilder[WebSocketFrame]): Future[RouteResult] = {
-    val sb = new StringBuilder(512)
-    WebsocketHandshake.serverHandshake(reqHeaders) match {
-      case Left((i, msg)) =>
-        logger.info(s"Invalid handshake: $i: $msg")
-        sb.append("HTTP/1.1 ").append(i).append(' ').append(msg).append('\r').append('\n')
-          .append('\r').append('\n')
-
-        channelWrite(ByteBuffer.wrap(sb.result().getBytes(StandardCharsets.ISO_8859_1)))
-          .map(_ => Close)
-
-      case Right(hdrs) =>
-        logger.info("Starting websocket request")
-        sb.append("HTTP/1.1 101 Switching Protocols\r\n")
-        hdrs.foreach { case (k, v) => sb.append(k).append(": ").append(v).append('\r').append('\n') }
-        sb.append('\r').append('\n')
-
-        // write the accept headers and reform the pipeline
-        channelWrite(ByteBuffer.wrap(sb.result().getBytes(StandardCharsets.ISO_8859_1))).map{ _ =>
-          logger.debug("Switching pipeline segments for upgrade")
-          val segment = wsBuilder.prepend(new WebSocketDecoder(false))
-          this.replaceTail(segment, true)
-          Upgrade
-        }
-    }
-  }
-
-  private def badRequest(msg: BadRequest): Unit = {
-    val sb = new StringBuilder(512)
-    sb.append("HTTP/").append(1).append('.')
-      .append(minor).append(' ').append(400)
-      .append(" Bad Request\r\n\r\n")
-
-    channelWrite(ByteBuffer.wrap(sb.result().getBytes(StandardCharsets.ISO_8859_1)))
-      .onComplete(_ => shutdownWithCommand(Cmd.Disconnect))
+  private def make5xx(error: Throwable): RouteAction = {
+    logger.error(error)("Failed to service request. Sending 500 response.")
+    RouteAction.String(500, "Internal Server Error", Nil, "Internal Server Error.")
```

Comment:
I'm not sure if you mean that it could be added to the `RouteAction` or not, but that seems like a good idea.